## Data 612- Project-4
### Santosh Manjrekar


## Recommender Systems Accuracy
1. As in your previous assignments, compare the accuracy of at least two recommender system algorithms against your offline data.
2. Implement support for at least one business or user experience goal such as increased serendipity, novelty, or diversity.
3. Compare and report on any change in accuracy before and after you’ve made the change in #2.
4. As part of your textual conclusion, discuss one or more additional experiments that could be performed and/or metrics that could be evaluated only if online evaluation was possible. Also, briefly propose how you would design a reasonable online evaluation environment.

### Let's evaluate SVD or matrix facorisation based model for RMSE and MAE

In [5]:
from surprise import SVD
from surprise import Dataset
from surprise.model_selection import cross_validate

# Load the movielens-100k dataset (download it if needed).
data = Dataset.load_builtin('ml-100k')

# Use the famous SVD algorithm.
algo = SVD()

# Run 5-fold cross-validation and print results.
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9364  0.9332  0.9437  0.9247  0.9381  0.9352  0.0063  
MAE (testset)     0.7379  0.7350  0.7445  0.7287  0.7392  0.7371  0.0052  
Fit time          23.45   21.38   20.00   18.18   18.51   20.30   1.94    
Test time         0.71    0.72    0.58    0.72    0.69    0.68    0.05    


{'test_rmse': array([0.93639338, 0.93323705, 0.94365088, 0.92465889, 0.93812681]),
 'test_mae': array([0.73789147, 0.73503652, 0.74453918, 0.72867257, 0.73921768]),
 'fit_time': (23.449915409088135,
  21.37979030609131,
  20.000617504119873,
  18.18253517150879,
  18.505729913711548),
 'test_time': (0.7109711170196533,
  0.7154655456542969,
  0.576002836227417,
  0.7180001735687256,
  0.6850981712341309)}

In [7]:
### Let's evaluate KNN based model for RMSE and MAE

In [6]:
# Use the famous KNN algorithm.
from surprise import KNNBasic
algo = KNNBasic()
# Run 5-fold cross-validation and print results.
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNBasic on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9819  0.9785  0.9716  0.9854  0.9750  0.9785  0.0049  
MAE (testset)     0.7746  0.7717  0.7699  0.7792  0.7676  0.7726  0.0040  
Fit time          1.00    1.08    1.03    1.08    1.04    1.04    0.03    
Test time         7.15    7.55    7.39    7.24    8.18    7.50    0.36    


{'test_rmse': array([0.98190324, 0.97846739, 0.97155039, 0.9854378 , 0.97497504]),
 'test_mae': array([0.77464747, 0.77172554, 0.76985174, 0.779182  , 0.76757721]),
 'fit_time': (0.9999983310699463,
  1.077998399734497,
  1.025998592376709,
  1.0789971351623535,
  1.0419719219207764),
 'test_time': (7.148000478744507,
  7.546000242233276,
  7.393002986907959,
  7.2409961223602295,
  8.17698621749878)}

The SVD model looks to be performing better with RMSE 0.93 better than KNN RMSE which is 0.97

Implement support for at least one business or user experience goal such as increased serendipity, novelty, 
or diversity.

Diversity can be represented as 1-Similarity. Diversity can be improved by proposing the less similar movies to
the user.
Novelty is kind of surprise suggestions to the user. We can achieve this by recommending less popular movies to
the user. One way is filtering out the recommendations given by the model and kicking out the popular movies out of
that recommedndations.
Both improving diversity and novelty will impact the accuracy in negative way. That means accuracy will get reduced.